### DISCUSSION TITLE
Ads Router

### DISCUSSION THREAD

**Discussion by JianxinChen** [fbarresi/SoftBeckhoff#6] [General]
Hello Federico,

This is JC Chen from US. 

I am working on the same effort to containerize the Beckhoff ADS client driver for my project. 

Currently, my ADS client code has to depend on the local TcSysSrv service running to communicate with the PLC Ads server. 

I want to run a AmsTcpIpRouter in the background to route the communication to the PLC Ads server directly in a docker container, but somehow, the AdsClient can't connect to the PLC Ads server. 

Have you figured out how to run AmsTcpIpRouter in a docker? or I totally misunderstand the Beckhoff ADS communication mechanism. 

Thanks for my clue.

JC

**Comment by fbarresi** [comment]
Hi JC Chen!

Thank you for getting in touch with me.

I guess yes. I think you basically have two way to do that. You may start the net core router application from beckhoff from nuget as backgroundservice or stand alone application and use a setting file for configuring the routes.
Or you can do in the same way I did for this project with reflection (and a bit of reverse engineering) and create and integrate in the code your own implementation of the router service where you can dynamically add or change routes.

Did you already had look in the code?



Best regards,

Federico

**Reply by JianxinChen** [reply]
Hi, Federico, 

Thanks for your quick reply.

I did review your code. 

My error is at the following step 4:  
 _hTTUplinkArray = _client.CreateVariableHandle(TTUplinkArray);  <-- can't create handle. 

I also noticed that the _client.ClientAddress has a different port number than the one "48898" given in appsettings.json. 

I am trying to test ADS client communication using AdsRouterWorker without starting TcSysSrv service locally, but I can't.

Thanks for any advice. 

JC

Here are what I want to do:
1.  AdsRouterWorker:
 public class AdsRouterWorker : BackgroundService
    {
        private readonly ILogger<AdsRouterWorker> _logger;
        private readonly IConfiguration _configuration;
        //private const string NetID = "5.85.197.194.1.1";
        //private const string IPAddress = "169.254.212.160";
        public AdsRouterWorker(ILogger<AdsRouterWorker> logger, IConfiguration configRoot)
        {
            _logger = logger;
            _configuration = configRoot;
        }
        /// <summary>
        /// Here is the link with the similar effort
        /// https://github.com/fbarresi/SoftBeckhoff/tree/e562460dd475424471db1276516d7bf59c95b245
        /// </summary>
        /// <param name="cancel"></param>
        /// <returns></returns>
        protected override async Task ExecuteAsync(CancellationToken cancel)
        {
            AmsTcpIpRouter router = new AmsTcpIpRouter(_logger, _configuration);
            //var netId = new AmsNetId(NetID);
            //AmsTcpIpRouter router = new AmsTcpIpRouter(new AmsNetId(NetID));
            //Route route = new Route("PLC_Port_0", netId, IPAddress);
            //router.AddRoute(route);
            await router.StartAsync(cancel); // Start the router
        }
    }

2.  define the AmsRouter in appsettings.json
{
  "Logging": {
    "LogLevel": {
      "Default": "Information",
      "Microsoft": "Warning",
      "Microsoft.Hosting.Lifetime": "Information"
    }
  },
  "AllowedHosts": "*",
  "Kestrel": {
    "EndpointDefaults": {
      "Protocols": "Http2"
    }
  },
  "AmsRouter": {
    "Name": "LocalSystem",
    "NetId": "172.18.77.225.1.1",
    "TcpPort": 48898,
    "RemoteConnections": [
      {
        "Name": "RemoteSystem",
        "Address": "169.254.212.160",
        "NetId": "5.85.197.194.1.1",
        "Type": "TCP_IP"
      }
    ]
  }
}

3.  Add AdsRouterWorker in Startup.cs
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddGrpc();
            services.AddSingleton<IDataAcqDriver, BeckoffAdsClientDriver>();
            services.AddHostedService<AdsRouterWorker>();
        }
4. Ads client connection
        public void Connect()
        {
            try
            {
                _client = new AdsClient();
                //"NetId": "172.18.77.225.1.1",
                //"TcpPort": 48898,
                _client.Connect("5.85.197.194.1.1", 851);   

                _hTTUplinkArray = _client.CreateVariableHandle(TTUplinkArray);
                _hTTUplinkFlag = _client.CreateVariableHandle(TTUplinkFlag);
                _hDownLinkArray = _client.CreateVariableHandle(DownLinkArray);
                _hStateUplinkArray = _client.CreateVariableHandle(StateUplinkArray);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Fail to connect PLC Net ID");
                Disconnect();
            }
        }

**Reply by JianxinChen** [reply]
By the way, the above code is running locally, not in a docker container yet.
